### DISCUSSION TITLE
Default commit-message for merge commits?

### DISCUSSION THREAD

**Discussion by BatmanAoD** [jj-vcs/jj#2973] [Q&A]
By default, when creating a merge commit, no description is created. This is somewhat unhelpful when that commit needs to be pushed to an upstream `git` repo, since commits without descriptions can't be pushed. Is there a way to configure a default message, preferably imitating Git's "merged branch [foo] into [bar]"?

**Comment by thoughtpolice** [comment]
Maybe we could have a single on/off configuration flag that means "do we automatically generate a message when we create commits with multiple parents." `core.auto-merge-message = true` or something?

Note that for the immediate moment, you should be able to do something like `jj new -m 'merge commit: foo + bar' r1 r2 r3...` to create a multi-parent merge and set the resulting message immediately.

**Reply by BatmanAoD** [reply]
Yes, an on/off config flag would be ideal. Thanks for the pointer about the positional args for `new` being parent commits; I haven't made any aliases yet but that would be a good one.

**Comment by amiryal** [comment]
[This article on LWN.net](https://lwn.net/Articles/926240/) is a good read on the subject of merge commit messages. Select quotes:

> […] Torvalds insists that each merge be properly explained. Merges are commits too, and their changelog should say why the merge is being done. It is easy to just accept the default message that Git creates when adding a merge commit, but the resulting commit will have no useful explanation, which is the equivalent of waving a large red flag for Torvalds to see. Unexplained merges thus have a high likelihood of generating one of those unwanted replies. 

> One way to determine whether a merge is needed is to explain the need for it; if the maintainer cannot write a changelog making the need for the merge clear, then perhaps the merge should not be done at all. 

> Take the time to write an actual commit message explaining why the merge was done. But, before that, take a moment to think about whether there actually _is_ a reason to do the merge. 

Whenever my teammates merge the trunk into their topic branch, I keep asking them why and quoting that last bit from the article.

So, going back on topic here, my point is that JJ has a better default than Git, since the automatic message in Git means more people neglect to _think_ about the message and replace it with a better one, thus making it almost pure noise. If the default is empty, then every non-empty message must signal something.

**Reply by amiryal** [reply]
The above holds well **when interop with Git is in mind**, since the tooling in Git does not do a good job of handling merge commits that also add changes besides a clean merge. Because of that, when you start new work in Git that requires merging some prerequisites first, you’d better create separate merge commits for the prerequisites, or the result will be hard to review in Git.

It’s worth pointing out, that JJ lets you start working immediately _on the merge commit itself_ without issues. You find your new work depends on fresh prerequisites from different branches? — rebase your work on top of all of them! The fact that your work is based on them will explain itself in the history graph.

Let’s say you started a line of work and the first commit is `A`. Now you find you depend on both `topicX` and `topicY`. In JJ, it is perfectly fine to `jj rebase -s A -d topicX -d topicY`. Notice that `A` itself is simply a diff with a description of the changes that it introduces, free to ignore the fact that it happens to be a “merge commit” right now.

If this work needs to be shared over Git with others who do not have JJ, then better separate the merge with a follow-up `jj new -B A -m 'Collect prerequisites' --no-edit`.

**Reply by BatmanAoD** [reply]
I don't really think Linus's rationale for needing to explain "why" every merge occurs makes sense for the typical practices I've seen for software maintained by a single team. In that context, branches are typically "feature" branches (focused on a single story or bugfix and rarely, if ever, long-lived), and there is almost never any reason to keep a feature branch behind the trunk branch. It's important that the code being reviewed and tested matches what will actually be in the trunk once a pull request is merged; that alone is sufficient reason to keep branches up to date. Yes, you can do so without a merge commit by doing rebases instead, but I have never believed that rebases are strictly preferable to merges.

I actually think that if all users had `jj` and the non-git backend was used for everything, it would be perfectly fine to leave commit messages empty for clean merges.

What does `new -B -A` actually do? It sounds like those should be mutually incompatible.

**Reply by amiryal** [reply]
A bad choice of placeholder rev name `A`, looks like `-A`. It is meant to insert a new rev before `A`, to replace it as the merge junction between `topicX`, `topicY` and the rest of the line of work.

**Reply by amiryal** [reply]
> It's important that the code being reviewed and tested matches what will actually be in the trunk once a pull request is merged

Reviewed — sure, reviewers may checkout the code with `jj new featureBranch 'trunk()'` in their own working copy and abandon it afterwards. Tested — yes, that is a good cause, but…

> that alone is sufficient reason to keep branches up to date.

No, that is the job of the CI pipeline, and there is no reason to keep that merge commit in the feature branch forever. At this point, I feel I am tilting at windmills (not the first time I am having this argument this week), so maybe I shouldn’t have said anything.

**Reply by BatmanAoD** [reply]
I'm sympathetic to what you're saying, I just don't think it applies to every project or every team. In any of these situations, I'd agree that merge commits should have a specific reason to exist:
* in a large, well-structured project with minimal improper coupling and a very good CI setup
* on a team that almost exclusively uses rebases rather than merges to update feature branches
* on a project like the Linux kernel where every maintainer's fork is long-lived and they are not necessarily kept in sync by default (i.e. when pulls are truly intentional "pulls" and not just the expected outcome of every unit of work)

....but in my professional career, I haven't worked on any projects like that, and I can't think of a single time when writing a specific commit message to explain why a merge is happening at all (as opposed to how conflicts were resolved or whether any related changes had to be made to fix build or test issues caused by the merge) would have been beneficial to me or anyone else.

**Reply by joyously** [reply]
> (as opposed to how conflicts were resolved or whether any related changes had to be made to fix build or test issues caused by the merge) 

I don't know that much about Git, but with Breezy, a merge is not a commit. That means you can resolve conflicts and test before doing the normal commit with a normal message about what is in it. I'm not sure how that applies to `jj` where everything is committed.

**Reply by BatmanAoD** [reply]
In git and jj, merges are commits.
